core/src/main/scala/slamdata/AutoBump.scala

```diff
@@ -0,0 +1,369 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package precog
+import java.io.File
+import java.nio.file.Files
+
+import org.http4s.Uri
+
+import cats.Monad
+import cats.data.EitherT
+import cats.effect.IO.contextShift
+import cats.effect.{ContextShift, IO, Sync}
+import cats.implicits._
+import github4s.GithubResponses.{GHException, GHResponse, GHResult}
+import github4s.domain._
+import precog.domain.{PullRequestDraft, PullRequestUpdate}
+import sbt.{Logger, url}
+import sbttrickle.metadata.OutdatedRepository
+import fs2.{Chunk, Stream}
+
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.sys.process._
+import scala.util.Try
+import scala.util.matching.Regex
+
+object AutoBump {
+  sealed trait Warnings {
+    def warn(log: Logger): IO[Unit]
+  }
+  case object UpdateError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("was unable to run `sbt update` following the trickle application")
+      log.warn("this may mean that the some artifacts are not yet propagated; skipping")
+    }
+  }
+  case object NoChangesError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("git-commit exited with error")
+      log.warn("this usually means the target repository was *already* at the latest version but hasn't published yet")
+      log.warn("you should check for a stuck trickle PR on that repository")
+    }
+  }
+  case object PushError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("git-push exited with error")
+      log.warn("this usually means some other repository updated the pull request before this one")
+    }
+  }
+  case class NotOldest(oldest: PullRequestDraft, draft: PullRequestDraft) extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn(s"pull request ${draft.number} is newer than existing pull request ${oldest.number}")
+      log.warn("this usually means two or more repositories finished build at the same time,")
+      log.warn("and some other repository beat this one to pull request creation.")
+    }
+  }
+
+  sealed abstract class ChangeLabel(val label: String)
+  object ChangeLabel {
+    val values: List[ChangeLabel] = List(RevisionChange, FeatureChange, BreakingChange)
+    val order: Map[ChangeLabel, Int] = values.zipWithIndex.toMap
+    implicit val ordering: Ordering[ChangeLabel] = Ordering.by(order)
+    val fromString: Map[String, ChangeLabel] = values.map(change => change.label -> change).toMap
+    val labelPattern: Regex = values.mkString("|").r
+
+    def apply(label: String): Option[ChangeLabel] = {
+      fromString.get(label)
+    }
+
+    def unapply(arg: String): Option[ChangeLabel] = labelPattern.findFirstIn(arg).flatMap(apply)
+  }
+  case object RevisionChange extends ChangeLabel("version: revision")
```

Comment:
One of the advantages to putting things into companion objects is you can simplify these names. Rather than `RevisionChange`, you'll get just `Revision`, inside of `ChangeLabel`.

Reply:
True. Curiously, nowhere in the code actually directly refers to these. I struggled on whether to make the code in `SbtPrecogBase` that decides what the label should be directly refer to these or not. I left this as is to avoid circular references, though both ADTs really ought to be moved out of `AutoBump`.